_source/_posts/2024-04-30-express-universal-logout.md

```diff
@@ -0,0 +1,674 @@
+  --url http://localhost:3333/global-token-revocation \
+  --header 'Content-Type: application/json' \
+  --data '{
+  "sub_id": {
+    "format": "email",
+    "email": "trinity@whiterabbit.fake"
+  }
+}'
+```
+
+You will get a 401 Unauthorized, meaning the authorization provided was invalid. This error is valid because we didn't send a token to our authorization headers. Let's fix that! We know from our database that the API key to our database is 131313. You can confirm this by opening another terminal and running `npx prisma studio`. A new window will open with a UI showing you the tables in your database; inspect the org table to see the column apikey for org 1. With that, go ahead and add this missing info to your cURL request. 
+
+It is crucial to authenticate with the correct API key as it specifies the org with which the user is associated. We'll also need to incorporate this information into the code. How might you do that? Thanks to the Passport.js library and how we configured the auth strategy in the `apps/api/src/main.ts`, we have information about the org the request is about, which is stored in `req['user']`. Try adding `console.log(req['user'])` to see what I mean. Furthermore, we can now access the org context with `req['user']['id']`. We'll need this to find the exact user within a specific org, and it will be beneficial if your app handles multitenancy. Add the following code to `apps/api/src/universalLogout.ts`: 
+
+```ts
+universalLogoutRoute.post('/global-token-revocation', async (req, res) => {
+  // 204 When the request is successful
+  const httpStatus = 204;
+
+  // 400 If the request is malformed
+  if (!req.body) {
+    res.status(400);
+  }
+  
+  // Make sure an API token is passed
+  const authHeaders = req.headers.authorization
+  if(!authHeaders){
+    return res.sendStatus(401)
+  }
+
+   // Find the user by email linked tothe org id associated with the API key provided
+  const domainOrgId = req['user']['id']
+  const newRequest:IRequestSchema = req.body;
+  const { email } = newRequest.sub_id;
+  const user = await prisma.user.findFirst({
+    where: {
+      email: email,
+      org: { id: domainOrgId } ,
+    },
+  });
+
+  // 404 User not found 
+  if (!user) {
+    res.sendStatus(404);
+  }
+
+universalLogoutRoute.use((err,req,res,next) => {
+  if(err){
+
+    return res.sendStatus(404)
+  }
+})
+```
+
+Moving right along, now that we have the target user of a specific org. Let's figure out how to target their application session and end it.
+
+## Implement the forced logout for a user
+
+In this section, we'll work towards ending a user's session and logging them out of the app.
+
+### End a user's session
+
+In Express, we'll need to access a user's session in the session store from the express-session library. Create a new file called sessionStore.ts under the `apps/api/src` folder and create a variable called store. A separate file to reference the `store` variable in multiple files. 
```

Comment:
```suggestion
In Express, we'll need to access a user's session in the session store from the `express-session` library. Create a new file called `sessionStore.ts` under the `apps/api/src` folder and create a variable named `store`. A separate file to reference the `store` variable in multiple files. 
```

I don't understand what the stand-alone sentence
> A separate file to reference file to reference the `store` variables in multiple files. 

means. I think it's missing something.

Reply:
addressed by adding the following sentences:

Notice we created a separate file to reference `store` in multiple files. For example, we'll need it in the `apps/api/src/main.ts` file and the `apps/api/src/universalLogout.ts` file.